streams/src/main/java/org/apache/kafka/streams/processor/internals/TransactionManager.java

```diff
@@ -0,0 +1,189 @@
+            try {
+                producer.initTransactions();
+            } catch (final TimeoutException exception) {
+                final String errorMessage = "Timeout exception caught when initializing transactions for task " + taskId + ". " +
+                    "\nThe broker is either slow or in bad state (like not having enough replicas) in responding to the request, " +
+                    "or the connection to broker was interrupted sending the request or receiving the response. " +
+                    "\n Consider overwriting `max.block.ms` to a larger value to avoid timeout errors";
+
+                // TODO K9113: we do NOT try to handle timeout exception here but throw it as a fatal error
+                throw new StreamsException(errorMessage, exception);
+            } catch (final KafkaException exception) {
+                throw new StreamsException("Error encountered while initializing transactions for task " + taskId, exception);
+            }
+        }
+    }
+
+    private void maybeBeginTransaction() throws ProducerFencedException {
+        if (eosAlphaEnabled && !transactionInFlight) {
+            try {
+                producer.beginTransaction();
+            } catch (final ProducerFencedException error) {
+                throw new TaskMigratedException(taskId, "Producer get fenced trying to begin a new transaction", error);
+            } catch (final KafkaException error) {
+                throw new StreamsException("Producer encounter unexpected error trying to begin a new transaction for task " + taskId, error);
+            }
+            transactionInFlight = true;
+        }
+    }
+
+    public Future<RecordMetadata> send(final ProducerRecord<byte[], byte[]> record,
+                                       final Callback callback) {
+        maybeBeginTransaction();
+        try {
+            return producer.send(record, callback);
+        } catch (final KafkaException uncaughtException) {
+            if (isRecoverable(uncaughtException)) {
+                // producer.send() call may throw a KafkaException which wraps a FencedException,
+                // in this case we should throw its wrapped inner cause so that it can be captured and re-wrapped as TaskMigrationException
+                throw new TaskMigratedException(taskId, "Producer cannot send records anymore since it got fenced", uncaughtException.getCause());
+            } else {
+                final String errorMessage = String.format(
+                    "Error encountered sending record to topic %s%s due to:%n%s",
+                    record.topic(),
+                    taskId == null ? "" : " for task " + taskId,
+                    uncaughtException.toString());
+                throw new StreamsException(errorMessage, uncaughtException);
+            }
+        }
+    }
+
+    private static boolean isRecoverable(final KafkaException uncaughtException) {
```

Comment:
nit: rename to `isProducerFenced`?

Reply:
An `UnknownProducerIdException` does not imply fencing? Does it?

Reply:
It was called `isRecoverable()` before -- think I just leave it.